src/target/riscv/riscv.c

```diff
@@ -4376,96 +4391,163 @@ static bool gdb_regno_cacheable(enum gdb_regno regno, bool is_write)
 }
 
 /**
- * This function is called when the debug user wants to change the value of a
- * register. The new value may be cached, and may not be written until the hart
- * is resumed. */
-int riscv_set_register(struct target *target, enum gdb_regno regid, riscv_reg_t value)
+ * This function is used internally by functions that change register values.
+ * If `write_through` is true, it is ensured that the value of the target's
+ * register is set to be equal to the `value` argument. The cached value is
+ * updated if the register is cacheable.
+ */
+static int riscv_set_or_write_register(struct target *target,
+		enum gdb_regno regid, riscv_reg_t value, bool write_through)
 {
 	RISCV_INFO(r);
-	LOG_DEBUG("[%s] %s <- %" PRIx64, target_name(target), gdb_regno_name(regid), value);
 	assert(r->set_register);
 
 	keep_alive();
 
-	/* TODO: Hack to deal with gdb that thinks these registers still exist. */
-	if (regid > GDB_REGNO_XPR15 && regid <= GDB_REGNO_XPR31 && value == 0 &&
-			riscv_supports_extension(target, 'E'))
-		return ERROR_OK;
+	if (!target->reg_cache) {
+		assert(!target_was_examined(target));
+		LOG_TARGET_DEBUG(target,
+				"No cache, writing to target: %s <- 0x%" PRIx64,
+				gdb_regno_name(regid), value);
+		return r->set_register(target, regid, value);
+	}
 
-	struct reg *reg = &target->reg_cache->reg_list[regid];
-	buf_set_u64(reg->value, 0, reg->size, value);
+	struct reg *reg = get_reg_cache_entry(target, regid);
 
-	if (gdb_regno_cacheable(regid, true)) {
-		reg->valid = true;
-		reg->dirty = true;
-	} else {
+	if (!reg->exist) {
+		LOG_TARGET_DEBUG(target, "Register %s does not exist.", reg->name);
+		return ERROR_FAIL;
+	}
+
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_DEBUG(target,
+				"Target not halted, writing to target: %s <- 0x%" PRIx64,
+				reg->name, value);
+		return r->set_register(target, regid, value);
+	}
+
+	const bool need_to_write = !reg->valid || reg->dirty ||
+		value != buf_get_u64(reg->value, 0, reg->size);
+	const bool cacheble = gdb_regno_cacheable(regid, need_to_write);
+
+	if (!cacheble || (write_through && need_to_write)) {
+		LOG_TARGET_DEBUG(target,
+				"Writing to target: %s <- 0x%" PRIx64 " (valid=%s, dirty=%s)",
+				reg->name, value, reg->valid ? "true" : "false",
+				reg->dirty ? "true" : "false");
```

Comment:
For easier troubleshooting, here you can also print whether the register is cacheable or not.

```suggestion
		LOG_TARGET_DEBUG(target,
				"Writing to target: %s <- 0x%" PRIx64 " (cacheable=%s, valid=%s, dirty=%s)",
				reg->name, value, cacheable ? "true" : "false", reg->valid ? "true" : "false",
				reg->dirty ? "true" : "false");
```

Reply:
Addressed.